plugins/blueprint.cpp

```diff
@@ -592,179 +612,200 @@ command_result do_transform(DFCoord start, DFCoord end, string name, uint32_t ph
+{
+    auto L = Lua::Core::State;
+    color_ostream_proxy out(Core::getInstance().getConsole());
+    Lua::StackUnwinder top(L);
+
+    if (!lua_checkstack(L, parameters.size() + 2) ||
+        !Lua::PushModulePublic(
+            out, L, "plugins.blueprint", "parse_commandline"))
+    {
+        out.printerr("Failed to load blueprint Lua code");
+        return false;
+    }
+
+    Lua::Push(L, &opts);
+
+    for (const string &param : parameters)
+        Lua::Push(L, param);
+
+    if (!Lua::SafeCall(out, L, parameters.size() + 1, 0))
+        return false;
+
+    return true;
+}
+
+static bool do_gui(const vector<string> &parameters)
+{
+    auto L = Lua::Core::State;
+    color_ostream_proxy out(Core::getInstance().getConsole());
+    Lua::StackUnwinder top(L);
+
+    if (!lua_checkstack(L, parameters.size() + 1) ||
+        !Lua::PushModulePublic(out, L, "plugins.blueprint", "do_gui"))
+    {
+        out.printerr("Failed to load blueprint Lua code");
+        return false;
+    }
+
+    for (const string &param : parameters)
+        Lua::Push(L, param);
+
+    if (!Lua::SafeCall(out, L, parameters.size(), 0))
+        return false;
+
+    return true;
+}
+
+static void print_help()
 {
-    return  find(parameters.begin(), parameters.end(), option) != parameters.end();
+    auto L = Lua::Core::State;
+    color_ostream_proxy out(Core::getInstance().getConsole());
+    Lua::StackUnwinder top(L);
+
+    if (!lua_checkstack(L, 1) ||
+        !Lua::PushModulePublic(out, L, "plugins.blueprint", "print_help") ||
+        !Lua::SafeCall(out, L, 0, 0))
+    {
+        out.printerr("Failed to load blueprint Lua code");
+    }
 }
 
 command_result blueprint(color_ostream &out, vector<string> &parameters)
 {
-    if (parameters.size() < 4 || parameters.size() > 8)
-        return help(out);
+    if (parameters.size() >= 1 && parameters[0] == "gui")
+    {
+        return do_gui(parameters) ? CR_OK : CR_FAILURE;
+    }
+
+    blueprint_options options;
+    if (!get_options(options, parameters) || options.help)
+    {
+        print_help();
+        return options.help ? CR_OK : CR_FAILURE;
+    }
+
     CoreSuspender suspend;
     if (!Maps::IsValid())
     {
         out.printerr("Map is not available!\n");
         return CR_FAILURE;
     }
-    int32_t x, y, z;
-    if (!Gui::getCursorCoords(x, y, z))
-    {
-        out.printerr("Can't get cursor coords! Make sure you have an active cursor in DF.\n");
-        return CR_FAILURE;
-    }
-    DFCoord start (x, y, z);
-    DFCoord end (x + stoi(parameters[0]), y + stoi(parameters[1]), z + stoi(parameters[2]));
-    uint32_t option = 0;
-    if (parameters.size() == 4)
+
+    // start coordinates can come from either the commandline or the map cursor
+    DFCoord start(options.start);
+    if (options.start.x == -30000)
     {
-        option = DIG | BUILD | PLACE | QUERY;
+        int32_t x, y, z;
+        if (!Gui::getCursorCoords(x, y, z))
```

Comment:
Could this be simplified to set fields of `start` directly? `getCursorCoords()` does modify its parameters even on failure, but you're returning immediately in that case anyway.
```suggestion
        if (!Gui::getCursorCoords(start.x, start.y, start.z))
```

Reply:
annoyingly, no (unless I do some casting). `start` is a `df::coord` which has `int16_t` fields. `getCursorCoords` takes `int32_t` params. I could cast, but I prefer to avoid that if possible. I've found that casually flirting with casting can cause strange and terrible bugs.

Reply:
Oh, thanks, I missed that. Casting references is just as dangerous as casting pointers, and if it compiled at all, it would end up clobbering memory since the function expects 4-byte parameters.

Maybe templating `getCursorCoords` or adding an overload for `df::coord` would be useful (but not required).

Reply:
ah, yeah, I can do that.
done.